### PR TITLE
NOTES.UNIX: mention more runtime search path flags.

### PR DESCRIPTION
**Pull request by dot-asm**
Also add "Linking your application" paragrapth.

### CODE REVIEW

**Comment by dot-asm** [comment]
It was suggested that AIX might deserve separate note or paragraph in NOTES.UNIX. Thing is that it was rather needed earlier, i.e. prior overhaul in #6487. Because it *was* [tricky](https://github.com/openssl/openssl/pull/6560#issuecomment-399674517) to do the right thing, and it was confusing, as it didn't follow vendor's pattern. But not anymore. Now AIX behaves pretty much as any other Unix, all it takes is straight-forward -L/some/where/lib -lssl -lcrypto (plus whatever developer should already know about AIX). In other words there is no unique OpenSSL-specific information that developer needs to know *anymore*. Hence there is no real need for AIX-specific note.

Well, there is one thing that is special about AIX, how to link with static libraries. But thing is that it's special even for other systems. Shall we add extra paragraph that discusses static link, or shall we leave it as exercise for user? But as an example, if you want to link with libcrypto statically on say Linux, you'd either have to specify it by full path name, or pass -Wl,-Bstatic,-lcrypto,-Bdynamic. While on AIX you'd have to -lcrypto_a.

**Comment by levitte** [comment]
I think that a section on linking with static libraries is desirable.

File: NOTES.UNIX [hunk 1]
```diff
@@ -89,3 +90,22 @@
...
+ It might be worth noting that some/most ELF systems implement support
+ for relative runtime search path, relative to directory containing
+ the executable object, by interpreting $ORIGIN along with some other
+ internal variables. Consult your system documentation.
+
+ Linking your application
+ ------------------------
+
+ Above is the story of building and installing OpenSSL shared libraries
+ and application. But third-party applications dynamically linked with
```
Comment (levitte) [inline NOTES.UNIX]: As mentioned in #6560, I don't think the first sentence is necessary at all (yes, I can see such a construct it a piece of literature, to create a sense of continuity between pieces of a story, but I don't think that applies here).  The whole paragraph can start with "Third party applications ..."

File: NOTES.UNIX [hunk 2]
```diff
@@ -89,3 +90,22 @@
 
     $ ./config --prefix=/usr/local/ssl --openssldir=/usr/local/ssl \
         '-Wl,--enable-new-dtags,-rpath,$(LIBRPATH)'
+
+ It might be worth noting that some/most ELF systems implement support
+ for relative runtime search path, relative to directory containing
```
Comment (levitte) [inline NOTES.UNIX]: Transported from #6560: s/for relative runtime search path, relative to directory containing/for runtime search path relative to directory containing/

**Comment by dot-asm** [comment]
Feedback addressed, [presumably].

File: NOTES.UNIX [hunk 3]
```diff
@@ -92,20 +92,26 @@
...
- -L/whatever/path -lssl -lcrypto. But there also are cases when you'd
- have to specify runtime search path yourself. Consult your system
- documentation and use above section as inspiration...
+ Third-party applications dynamically linked with OpenSSL (or any other)
+ shared library face exactly same problem with non-default locations.
+ Abovementioned OpenSSL config options might or might not have bearing
+ on linking of the target application. "Might" means that under some
+ circumstances it would be sufficient to link with OpenSSL shared library
+ "naturally", i.e. with -L/whatever/path -lssl -lcrypto. But there also
+ are cases when you'd have to explicitly specify runtime search path
```
Comment (levitte) [inline NOTES.UNIX]: s/But the also are cases/But there are also cases/

File: NOTES.UNIX [hunk 4]
```diff
@@ -92,20 +92,26 @@
...
+ when linking your applcation. Consult your system documentation and use
+ above section as inspiration...
+
+ Shared OpenSSL build installs even static libraries. Linking with the
```
Comment (levitte) [inline NOTES.UNIX]: s/Shared OpenSSL build installs even static libraries/Shared OpenSSL builds also install static libraries/

File: NOTES.UNIX [hunk 5]
```diff
@@ -92,20 +92,26 @@
...
+ latter is likely to require special care, because customarily linkers
+ look for shared libraries first, and tend to remain "blind" to static
```
Comment (levitte) [inline NOTES.UNIX]: s/because customarily linkers look/because linkers usually look/

File: NOTES.UNIX [hunk 6]
```diff
@@ -92,20 +92,26 @@
...
+ OpenSSL libraries. Referring to system documentation would suffice, if
+ not for a corner case. On AIX static libraries [in shared build] are
```
Comment (levitte) [inline NOTES.UNIX]: In natural languages, use ( and ) rather than [ and ]

File: NOTES.UNIX (hunk 3 above)
Comment (dot-asm) [reply NOTES.UNIX]: It says "but there are also cases"... Or what am I missing?

Comment (levitte) [reply NOTES.UNIX]: Nope, it says "But there also are cases".  "also" and "are" need to trade places.

Comment (dot-asm) [reply NOTES.UNIX]: Ah! I was hooked on "the" vs. "there"...

**Comment by dot-asm** [comment]
Next round...

File: NOTES.UNIX [hunk 7]
```diff
@@ -89,3 +90,28 @@
...
+ It might be worth noting that some/most ELF systems implement support
+ for runtime search path relative to the directory containing current
+ executable, by interpreting $ORIGIN along with some other internal
+ variables. Consult your system documentation.
+
+ Linking your application
+ ------------------------
+
+ Third-party applications dynamically linked with OpenSSL (or any other)
+ shared library face exactly same problem with non-default locations.
```
Comment (levitte) [inline NOTES.UNIX]: *the* same problem

(i.e. "the" missing)

File: NOTES.UNIX [hunk 8]
```diff
@@ -89,3 +90,28 @@
...
+ Abovementioned OpenSSL config options might or might not have bearing
```
Comment (levitte) [inline NOTES.UNIX]: s/Abovementioned OpenSSL config options/The OpenSSL config options mentioned above/

**Pull request closed without merging**